plaid/src/apis/okta/mod.rs

```diff
@@ -30,15 +41,103 @@ pub struct Okta {
 pub enum OktaError {
     BadData(FromUtf8Error),
     UnexpectedStatusCode(u16),
+    AuthenticationFailure,
+    BadPrivateKey,
+    JwtSignatureFailure,
+    BadJsonResponse
+}
+
+pub enum OktaOperation {
+    GetUserInfo,
+    RemoveUserFromGroup
+}
+
+impl OktaOperation {
+    /// Get the correct Okta OAuth 2.0 scope, depending on which operation needs to be performed.
+    pub fn to_okta_scope(&self) -> &str {
+        match self {
+            // https://developer.okta.com/docs/api/openapi/okta-management/management/tag/User/#tag/User/operation/getUser
+            Self::GetUserInfo => "okta.users.read",
+            // https://developer.okta.com/docs/api/openapi/okta-management/management/tag/Group/#tag/Group/operation/unassignUserFromGroup
+            Self::RemoveUserFromGroup => "okta.groups.manage"
+        }
+    }
 }
 
 impl Okta {
     pub fn new(config: OktaConfig) -> Self {
         let client = reqwest::Client::builder()
-            .timeout(Duration::from_secs(config.api_timeout_seconds))
+            .timeout(std::time::Duration::from_secs(config.api_timeout_seconds))
             .build()
             .unwrap();
 
         Self { config, client }
     }
+
+    /// Return an appropriate authorization header, to be used when making a REST call.
+    pub async fn get_authorization_header(&self, op: &OktaOperation) -> Result<String, OktaError> {
+        match &self.config.authentication {
+            Authentication::ApiKey { token } => Ok(format!("SSWS {}", token)),
+            Authentication::OktaApp { .. } => {
+                let access_token = self.get_access_token(op).await?;
+                Ok(format!("Bearer {}", access_token))
+            }
+        }
+    }
+
+    /// Construct a JWT signed with Plaid's private key, to be later exchanged for an access token.
+    fn get_jwt(&self) -> Result<String, OktaError> {
+        match &self.config.authentication {
+            Authentication::ApiKey { token: _ } => panic!("Trying to get a JWT but this is not configured as an Okta app. This should be impossible."),
```

Comment:
There is no way for a rule to make this happen correct?

Reply:
Correct, this is impossible because it all comes down from calling `get_access_token`, and that's only called if we authenticate as an Okta app. Let me slightly rearrange the code, because I think we can get rid of this.